presto-common/src/main/java/com/facebook/presto/common/block/BlockUtil.java

```diff
@@ -0,0 +1,286 @@
+ */
+package com.facebook.presto.common.block;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static java.lang.Math.ceil;
+import static java.lang.Math.max;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public final class BlockUtil
+{
+    private static final double BLOCK_RESET_SKEW = 1.25;
+
+    private static final int DEFAULT_CAPACITY = 64;
+    // See java.util.ArrayList for an explanation
+    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    private BlockUtil()
+    {
+    }
+
+    static void checkArrayRange(int[] array, int offset, int length)
+    {
+        requireNonNull(array, "array is null");
+        if (offset < 0 || length < 0 || offset + length > array.length) {
+            throw new IndexOutOfBoundsException(format("Invalid offset %s and length %s in array with %s elements", offset, length, array.length));
+        }
+    }
+
+    static void checkArrayRange(boolean[] array, int offset, int length)
+    {
+        requireNonNull(array, "array is null");
+        if (offset < 0 || length < 0 || offset + length > array.length) {
+            throw new IndexOutOfBoundsException(format("Invalid offset %s and length %s in array with %s elements", offset, length, array.length));
+        }
+    }
+
+    static void checkValidRegion(int positionCount, int positionOffset, int length)
+    {
+        if (positionOffset < 0 || length < 0 || positionOffset + length > positionCount) {
+            throw new IndexOutOfBoundsException(format("Invalid position %s and length %s in block with %s positions", positionOffset, length, positionCount));
+        }
+    }
+
+    static void checkValidPositions(boolean[] positions, int positionCount)
+    {
+        if (positions.length != positionCount) {
+            throw new IllegalArgumentException(format("Invalid positions array size %d, actual position count is %d", positions.length, positionCount));
+        }
+    }
+
+    static void checkValidPosition(int position, int positionCount)
+    {
+        if (position < 0 || position >= positionCount) {
+            throw new IllegalArgumentException(format("Invalid position %s in block with %s positions", position, positionCount));
+        }
+    }
+
+    static int calculateNewArraySize(int currentSize)
+    {
+        // grow array by 50%
+        long newSize = (long) currentSize + (currentSize >> 1);
+
+        // verify new size is within reasonable bounds
+        if (newSize < DEFAULT_CAPACITY) {
+            newSize = DEFAULT_CAPACITY;
+        }
+        else if (newSize > MAX_ARRAY_SIZE) {
+            newSize = MAX_ARRAY_SIZE;
+            if (newSize == currentSize) {
+                throw new IllegalArgumentException(format("Can not grow array beyond '%s'", MAX_ARRAY_SIZE));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: MAX_ARRAY_SIZE. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 91. Consider formatting this data with &#39;%d&#39; instead.